Madison-Taylor Fortich  
9291000361

### 1. Preprocessing the data:
a. Import the libraries and datasets    
b. Use ‘re’ library to extract the characters [a-zA-Z]  
c. Use ‘nltk’ library to extract to stem the sentence and remove the stop words.   
d. Leverage countvectorizer to vectorize the data.  
e. Split the train and test set.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv(r'Restaurant_Reviews.tsv', sep='\t')
data.head()

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1

In [3]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

corpus = []
for i in range(0,1000):
    review = re.sub('[^a-zA-Z]', ' ', data['Review'][i])
    review = review.lower()
    review = review.split()
    ps = PorterStemmer()
    all_stopwords = stopwords.words('english')
    all_stopwords.remove('not')
    review = [ps.stem(word) for word in review if not word in set(all_stopwords)]
    review = ' '.join(review)
    corpus.append(review)

[nltk_data] Downloading package stopwords to /Users/madi/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [4]:
# only printing the first 10
corpus[:10]

['wow love place',
 'crust not good',
 'not tasti textur nasti',
 'stop late may bank holiday rick steve recommend love',
 'select menu great price',
 'get angri want damn pho',
 'honeslti tast fresh',
 'potato like rubber could tell made ahead time kept warmer',
 'fri great',
 'great touch']

In [5]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=1500)
X = cv.fit_transform(corpus).toarray()
y = data.iloc[:, -1].values

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.20, random_state = 0)

In [7]:
print(X_train, X_test, y_train, y_test)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]] [[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]] [1 1 1 0 1 0 1 0 0 0 1 0 0 1 1 0 1 1 1 1 0 1 0 0 1 1 1 0 0 1 1 1 1 0 1 0 1
 1 1 0 1 1 1 1 0 1 0 0 1 0 0 1 1 0 0 0 1 1 0 1 0 1 1 0 0 1 1 0 1 0 0 0 1 1
 0 1 1 1 0 1 0 0 1 1 1 0 1 0 1 0 0 0 1 0 0 0 0 0 1 1 1 0 1 1 1 1 0 0 1 0 0
 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 1 1 1 1 1 0 1 0 1 0 1 0 0 1 0 1 0 0 1 0 1 0
 0 0 1 1 1 0 0 0 1 1 1 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 1 1 1 1 1 1 0
 0 1 0 1 1 1 1 0 0 1 0 0 0 0 0 1 1 1 0 1 1 0 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1
 0 0 0 0 0 1 0 1 0 0 1 0 1 0 1 1 0 1 1 1 0 0 1 1 1 1 1 0 1 1 0 0 0 1 0 0 1
 0 0 0 1 0 0 0 0 1 1 1 0 0 1 0 1 1 0 0 1 0 1 0 0 1 0 1 1 0 1 0 1 1 0 1 1 0
 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 0 1 0 1 0 1 1 1 0 1 0 1 1 1 1 1 1 1 0 0
 1 1 1 0 0 1 1 0 1 0 1 0 0 0 1 1 1 0 1 1 1 1 1 1 0 0 1 0 1 0 0 1 0 0 0 1 0
 0 1 1 1 1

### 2. Build the models: 
a. Build Logistic Regression Classifier, Naive Bayes, Decision Tree Classifier, Random Forest Classifier.**   
b. Use Grid Search CV to find the best parameters of each model, and train the models with the best parameters.**   
c. Predict with the test set**  
  
### 3. Evaluation: 
a. Calculate the classification_report and confusion matrix of each model.  
b. Write a brief report evaluating those models.  
  


**Logistic Regression**

In [8]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

classifier_lr = LogisticRegression(random_state=0)
classifier_lr.fit(X_train, y_train)
y_pred_lr = classifier_lr.predict(X_test)

In [9]:
from sklearn.model_selection import GridSearchCV

lr_params = {'C': [0.1, 1, 10, 100], 'penalty': ['l2'], 'solver': ['lbfgs', 'liblinear']}

grid_search_lr = GridSearchCV(estimator=classifier_lr, param_grid=lr_params, scoring='accuracy', cv=10, n_jobs=-1)
grid_search_lr.fit(X_train, y_train)

best_lr = grid_search_lr.best_estimator_
y_pred_lr_gs = best_lr.predict(X_test)

print("Best Parameters for Logistic Regression:", grid_search_lr.best_params_)
print("Test Set Prediction:", y_pred_lr_gs)

Best Parameters for Logistic Regression: {'C': 1, 'penalty': 'l2', 'solver': 'lbfgs'}
Test Set Prediction: [0 0 0 0 0 0 1 0 0 1 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0 1 0 0
 0 0 1 1 1 1 0 0 0 1 1 1 0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 0 0 1 0 1 0 1 1 1 0
 0 0 0 1 0 1 0 1 1 0 1 1 1 0 1 1 0 1 0 0 0 0 0 0 0 1 0 1 1 1 1 0 1 1 0 0 0
 1 0 1 0 1 1 1 1 1 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 1 0 0 1 1 0 0 1 0 1 0 0 0 1 1 1 0 0 0 0 0 0 1 1 0 0 0 0 1 0 1 1 1 0 0 0
 0 0 1 0 1 1 0 0 0 1 0 1 1 0 1]


In [10]:
from sklearn.metrics import classification_report, confusion_matrix

print("Logistic Regression - Classification Report")
print(classification_report(y_test, y_pred_lr))

print("Logistic Regression - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_lr))

print("Logistic Regression Grid Search - Classification Report")
print(classification_report(y_test, y_pred_lr_gs))

print("Logistic Regression Grid Search - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_lr_gs))

Logistic Regression - Classification Report
              precision    recall  f1-score   support

           0       0.74      0.82      0.78        97
           1       0.82      0.73      0.77       103

    accuracy                           0.78       200
   macro avg       0.78      0.78      0.77       200
weighted avg       0.78      0.78      0.77       200

Logistic Regression - Confusion Matrix
[[80 17]
 [28 75]]
Logistic Regression Grid Search - Classification Report
              precision    recall  f1-score   support

           0       0.74      0.82      0.78        97
           1       0.82      0.73      0.77       103

    accuracy                           0.78       200
   macro avg       0.78      0.78      0.77       200
weighted avg       0.78      0.78      0.77       200

Logistic Regression Grid Search - Confusion Matrix
[[80 17]
 [28 75]]


-  Grid Search did not improve the Logistic Regression model. The accuracy remained the same at 78% as well as the classification report. 

**Naive Bayes**

In [11]:
from sklearn.naive_bayes import GaussianNB

classifier_nb = GaussianNB()
classifier_nb.fit(X_train, y_train)
y_pred_nb = classifier_nb.predict(X_test)

In [12]:
nb_params = {'var_smoothing': [1e-9, 1e-8, 1e-7, 1e-6, 1e-5]}

grid_search_nb = GridSearchCV(estimator=classifier_nb, param_grid=nb_params, scoring='accuracy', cv=10, n_jobs=-1)
grid_search_nb.fit(X_train, y_train)

best_nb = grid_search_nb.best_estimator_
y_pred_nb_gs = best_nb.predict(X_test)

print("Best Parameters (Naive Bayes):", grid_search_nb.best_params_)
print("Test Set Prediction:", y_pred_nb_gs)

Best Parameters (Naive Bayes): {'var_smoothing': 1e-05}
Test Set Prediction: [1 1 1 0 0 1 1 0 1 1 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 1 1 1 0 1 1 1 1 1 0 1
 0 1 1 1 1 1 0 0 0 1 1 0 0 1 1 1 1 1 0 1 1 0 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1
 0 1 1 0 0 1 0 1 1 0 1 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 0 1 1 1 0 1 1 1 0 0 0
 1 0 1 1 0 1 1 1 1 1 0 1 1 0 0 1 1 0 1 1 1 0 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1
 1 1 1 0 1 1 1 0 1 1 1 1 1 0 0 1 0 0 1 0 0 0 0 1 1 0 0 1 0 1 0 0 1 0 0 1 0
 1 0 1 0 1 1 0 1 1 1 0 1 1 1 1]


In [13]:
print("Naive Bayes - Classification Report")
print(classification_report(y_test, y_pred_nb))

print("Naive Bayes - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_nb))

print("Naive Bayes Grid Search - Classification Report")
print(classification_report(y_test, y_pred_nb_gs))

print("Naive Bayes Grid Search - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_nb_gs))

Naive Bayes - Classification Report
              precision    recall  f1-score   support

           0       0.82      0.57      0.67        97
           1       0.68      0.88      0.77       103

    accuracy                           0.73       200
   macro avg       0.75      0.73      0.72       200
weighted avg       0.75      0.73      0.72       200

Naive Bayes - Confusion Matrix
[[55 42]
 [12 91]]
Naive Bayes Grid Search - Classification Report
              precision    recall  f1-score   support

           0       0.82      0.58      0.68        97
           1       0.69      0.88      0.77       103

    accuracy                           0.73       200
   macro avg       0.76      0.73      0.73       200
weighted avg       0.75      0.73      0.73       200

Naive Bayes Grid Search - Confusion Matrix
[[56 41]
 [12 91]]


- Grid Search improved the Naive Bayes model by a small margin. The accuracy is the same at 73%, but the True Negatives and False Positivws improved. 

**Decision Tree**

In [14]:
from sklearn.tree import DecisionTreeClassifier
classifier_tree = DecisionTreeClassifier(random_state=0)
classifier_tree.fit(X_train, y_train)
y_pred_tree = classifier_tree.predict(X_test)

In [15]:
tree_params = {'criterion': ['gini', 'entropy'], 'max_depth': [10, 20, 30, None], 'min_samples_split': [2, 10, 20]}

grid_search_tree = GridSearchCV(estimator=classifier_tree, param_grid=tree_params, scoring='accuracy', cv=10, n_jobs=-1)
grid_search_tree.fit(X_train, y_train)

best_tree = grid_search_tree.best_estimator_
y_pred_tree_gs = best_tree.predict(X_test)

print("Best Parameters (Decision Tree):", grid_search_tree.best_params_)
print("Test Set Prediction:", y_pred_tree_gs)

Best Parameters (Decision Tree): {'criterion': 'gini', 'max_depth': None, 'min_samples_split': 10}
Test Set Prediction: [0 0 0 0 0 1 1 1 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 0 0 1 0 1 1 0 0 1 0 0 1 1 0
 0 0 1 1 1 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 0 1 0 1 0 0 0 0 1 1 1 1 0
 0 0 0 1 0 0 0 1 0 0 1 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 1 0 0 0
 1 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 0 0 1 0 1 0 0 1 1 1 1 1 0 1 1 1 0 0 0 1 0
 0 0 0 1 1 1 0 0 1 0 1 0 0 0 1 1 1 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 1 1 0 0 0
 0 0 1 0 1 1 0 0 0 1 0 1 1 0 0]


In [16]:
print("Decision Tree - Classification Report")
print(classification_report(y_test, y_pred_tree))

print("Decision Tree - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_tree))

print("Decision Tree Grid Search - Classification Report")
print(classification_report(y_test, y_pred_tree_gs))

print("Decision Tree Grid Search - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_tree_gs))

Decision Tree - Classification Report
              precision    recall  f1-score   support

           0       0.70      0.79      0.74        97
           1       0.78      0.68      0.73       103

    accuracy                           0.73       200
   macro avg       0.74      0.74      0.73       200
weighted avg       0.74      0.73      0.73       200

Decision Tree - Confusion Matrix
[[77 20]
 [33 70]]
Decision Tree Grid Search - Classification Report
              precision    recall  f1-score   support

           0       0.70      0.84      0.76        97
           1       0.81      0.66      0.73       103

    accuracy                           0.74       200
   macro avg       0.75      0.75      0.74       200
weighted avg       0.76      0.74      0.74       200

Decision Tree Grid Search - Confusion Matrix
[[81 16]
 [35 68]]


- Grid Search improved the Decision Tree model by 1% in accuracy (73 to 74 percent). True Negatives and false positives also improved. Overall, there is some imporvement, but not by a lot.

**Random Forest**

In [17]:
from sklearn.ensemble import RandomForestClassifier

classifier_rf = RandomForestClassifier(random_state=0)
classifier_rf.fit(X_train, y_train)
y_pred_rf = classifier_rf.predict(X_test)

In [18]:
rf_params = {'n_estimators': [10, 50, 100, 200], 'criterion': ['gini', 'entropy'], 'max_depth': [10, 20, 30, None], 
             'min_samples_split': [2, 10, 20]}

grid_search_rf = GridSearchCV(estimator=classifier_rf, param_grid=rf_params, scoring='accuracy', cv=10, n_jobs=-1)
grid_search_rf.fit(X_train, y_train)

best_rf = grid_search_rf.best_estimator_
y_pred_rf_gs = best_rf.predict(X_test)

print("Best Parameters (Random Forest):", grid_search_rf.best_params_)
print("Test Set Prediction:", y_pred_rf_gs)

Best Parameters (Random Forest): {'criterion': 'gini', 'max_depth': 20, 'min_samples_split': 10, 'n_estimators': 200}
Test Set Prediction: [0 0 0 0 0 0 1 0 0 1 1 1 0 1 1 1 0 0 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 1 0 0
 0 0 1 1 0 0 0 0 0 0 1 0 0 1 0 1 1 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 1 1 0
 0 0 0 1 0 0 0 1 0 0 1 1 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 1 0 0 0
 0 0 1 0 0 1 0 1 1 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 1 0 1 1 1 0 0 0 0 0
 0 0 0 1 1 1 0 0 1 0 1 0 0 0 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 1 1 0 0 0 0
 0 0 1 0 1 1 0 0 0 0 0 1 1 0 0]


In [19]:
print("Random Forest - Classification Report")
print(classification_report(y_test, y_pred_rf))

print("Random Forest - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_rf))

print("Random Forest Grid Search - Classification Report")
print(classification_report(y_test, y_pred_rf_gs))

print("Random Forest Grid Search - Confusion Matrix")
print(confusion_matrix(y_test, y_pred_rf_gs))

Random Forest - Classification Report
              precision    recall  f1-score   support

           0       0.70      0.94      0.80        97
           1       0.91      0.62      0.74       103

    accuracy                           0.78       200
   macro avg       0.81      0.78      0.77       200
weighted avg       0.81      0.78      0.77       200

Random Forest - Confusion Matrix
[[91  6]
 [39 64]]
Random Forest Grid Search - Classification Report
              precision    recall  f1-score   support

           0       0.69      0.94      0.79        97
           1       0.91      0.60      0.73       103

    accuracy                           0.77       200
   macro avg       0.80      0.77      0.76       200
weighted avg       0.80      0.77      0.76       200

Random Forest Grid Search - Confusion Matrix
[[91  6]
 [41 62]]


- Grid Search didn't improve the Random Forest model. The accuracy decreased from 78 to 77%. There was also a decrease in precision, recall, and f-1 scores. 